File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/domene/statistikk/ÅrstallOgKvartal.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.domene.statistikk;

import lombok.AllArgsConstructor;
import lombok.Data;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

@Data
@AllArgsConstructor
public class ÅrstallOgKvartal implements Comparable<ÅrstallOgKvartal> {
    private final int årstall;
    private final int kvartal;


    public ÅrstallOgKvartal minusKvartaler(int antallKvartaler) {
        if (antallKvartaler < 0) {
            return plussKvartaler(-antallKvartaler);
        }
        ÅrstallOgKvartal årstallOgKvartal = new ÅrstallOgKvartal(årstall, kvartal);
        for (int i = 0; i < antallKvartaler; i++) {
            årstallOgKvartal = årstallOgKvartal.hentForrigeKvartal();
        }
        return årstallOgKvartal;
    }

    public ÅrstallOgKvartal plussKvartaler(int antallKvartaler) {
        if (antallKvartaler < 0) {
            return minusKvartaler(-antallKvartaler);
        }
        ÅrstallOgKvartal årstallOgKvartal = new ÅrstallOgKvartal(årstall, kvartal);
        for (int i = 0; i < antallKvartaler; i++) {
            årstallOgKvartal = årstallOgKvartal.hentNesteKvartal();
        }
        return årstallOgKvartal;
    }

    public static List<ÅrstallOgKvartal> range(ÅrstallOgKvartal fra, ÅrstallOgKvartal til) {
        List<ÅrstallOgKvartal> årstallOgKvartal = new ArrayList<>();
        for (ÅrstallOgKvartal i = fra; i.compareTo(til) <= 0; i = i.plussKvartaler(1)) {
            årstallOgKvartal.add(i);
        }
        return årstallOgKvartal;
    }

    private ÅrstallOgKvartal hentForrigeKvartal() {
        if (kvartal == 1) {
            return new ÅrstallOgKvartal(årstall - 1, 4);
        } else {
            return new ÅrstallOgKvartal(årstall, kvartal - 1);
        }
    }

    private ÅrstallOgKvartal hentNesteKvartal() {
        if (kvartal == 4) {
            return new ÅrstallOgKvartal(årstall + 1, 1);
        } else {
            return new ÅrstallOgKvartal(årstall, kvartal + 1);
        }
    }

    @Override
    public int compareTo(ÅrstallOgKvartal årstallOgKvartal) {
        return Comparator.comparing(ÅrstallOgKvartal::getÅrstall)
                .thenComparing(ÅrstallOgKvartal::getKvartal)
                .compare(this, årstallOgKvartal);
    }
}

File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/applikasjon/domenemodeller/ÅrstallOgKvartal.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller

data class ÅrstallOgKvartal(
    @get:JvmName("getÅrstall") val årstall: Int,
    val kvartal: Int
) : Comparable<ÅrstallOgKvartal> {

    init {
        require(!(kvartal > 4 || kvartal < 1)) { "Kvartal må være 1, 2, 3 eller 4" }
    }

    fun minusKvartaler(antallKvartaler: Int): ÅrstallOgKvartal {
        if (antallKvartaler < 0) {
            return plussKvartaler(-antallKvartaler)
        }
        var årstallOgKvartal = ÅrstallOgKvartal(årstall, kvartal)
        for (i in 0 until antallKvartaler) {
            årstallOgKvartal = årstallOgKvartal.forrigeKvartal()
        }
        return årstallOgKvartal
    }

    fun minusEttÅr(): ÅrstallOgKvartal {
        return ÅrstallOgKvartal(årstall - 1, kvartal)
    }

    fun plussKvartaler(antallKvartaler: Int): ÅrstallOgKvartal {
        if (antallKvartaler < 0) {
            return minusKvartaler(-antallKvartaler)
        }
        var årstallOgKvartal = ÅrstallOgKvartal(årstall, kvartal)
        for (i in 0 until antallKvartaler) {
            årstallOgKvartal = årstallOgKvartal.nesteKvartal()
        }
        return årstallOgKvartal
    }

    private fun forrigeKvartal(): ÅrstallOgKvartal {
        return if (kvartal == 1) {
            ÅrstallOgKvartal(årstall - 1, 4)
        } else {
            ÅrstallOgKvartal(årstall, kvartal - 1)
        }
    }

    private fun nesteKvartal(): ÅrstallOgKvartal {
        return if (kvartal == 4) {
            ÅrstallOgKvartal(årstall + 1, 1)
        } else {
            ÅrstallOgKvartal(årstall, kvartal + 1)
        }
    }

    override fun compareTo(other: ÅrstallOgKvartal): Int {
        return Comparator.comparing<ÅrstallOgKvartal, Int> { it.årstall }
            .thenComparing<Int> { it.kvartal }
            .compare(this, other)
    }

    operator fun rangeTo(other: ÅrstallOgKvartal): KvartalRange {
        return KvartalRange(this, other)
    }

    override fun toString(): String {
        return "$kvartal. kvartal $årstall"
    }

    companion object {
        @JvmStatic
        fun sisteFireKvartaler(årstallOgKvartal: ÅrstallOgKvartal): List<ÅrstallOgKvartal> =
            (0..3).map(årstallOgKvartal::minusKvartaler).sorted()

        @JvmStatic
        fun range(fra: ÅrstallOgKvartal, til: ÅrstallOgKvartal): List<ÅrstallOgKvartal> {
            val årstallOgKvartal: MutableList<ÅrstallOgKvartal> = ArrayList()
            var i = fra
            while (i <= til) {
                årstallOgKvartal.add(i)
                i = i.plussKvartaler(1)
            }
            return årstallOgKvartal
        }
    }
}


class KvartalRange(override val start: ÅrstallOgKvartal, override val endInclusive: ÅrstallOgKvartal) :
    ClosedRange<ÅrstallOgKvartal> {
    operator fun iterator(): Iterator<ÅrstallOgKvartal> {
        return KvartalIterator(start, endInclusive)
    }
}

class KvartalIterator(start: ÅrstallOgKvartal, private val endInclusive: ÅrstallOgKvartal) : Iterator<ÅrstallOgKvartal> {
    private var previous = start.copy()
    override fun hasNext() = previous <= endInclusive

    override fun next(): ÅrstallOgKvartal = previous.plussKvartaler(1)
}
